user-bag-service/src/main/kotlin/io/baum/userbagservice/users/web/UserService.kt

```diff
@@ -0,0 +1,35 @@
+package io.baum.userbagservice.users.web
+
+import arrow.core.Either
+import io.baum.userbagservice.users.error.RecordNotFoundException
+import io.baum.userbagservice.users.error.UnauthorisedUserException
+import io.baum.userbagservice.users.kafka.UserProducer
+import io.baum.userbagservice.users.kafka.UserRecord
+import io.baum.userbagservice.users.kafka.UserRepository
+import io.baum.userbagservice.users.tools.PasswordAuthenticator
+import io.baum.userbagservice.users.web.model.RemoteAddress
+import io.baum.userbagservice.users.web.model.UserModel
+import org.springframework.stereotype.Service
+import io.baum.userbagservice.users.tools.toDomain
+import io.baum.userbagservice.users.tools.toKafkaRecord
+
+@Service
+class UserService(
+        private val userProducer: UserProducer,
+        private val userRepository: UserRepository,
+        private val passwordAuthenticator: PasswordAuthenticator
+) {
+    fun createUser(user: UserModel) =
+        userProducer.publish(user.id, user.toKafkaRecord())
+
+    fun getUserById(id: String, password: String): Either<RemoteAddress, UserModel> =
+        userRepository.getUserById(id)
+                .map { record -> record ?: throw RecordNotFoundException(id) }
+                .map { record -> if(passwordAuthenticator.verifyPassword(password, record.password)) record.toDomain() else throw UnauthorisedUserException(id) }
+
+    fun getAllUsers(remote: Boolean): List<UserModel?> = if(!remote) {
```

Comment:
if `remote=true` it calls a method called `getAllUsersLocal` which seems a bit strange

Reply:
It does sound strange but the way it works is that during a get all users, we technically have a master - slave architecture in which the original instance the request was made to, the master, returns all it's own data and queries other instances for their data. The slaves, the 'remote' instances are to only return their own data without querying the other instances as obviously having each slave call the other instances would cause deadlock

Reply:
Code should be intuitive though where possible